README.md
# ArchiSteamFarm
[![Build Status](https://travis-ci.org/T-Wong/asf.svg?branch=master)](https://travis-ci.org/T-Wong/asf) [![Cookbook Version](https://img.shields.io/cookbook/v/asf.svg)](https://supermarket.chef.io/cookbooks/asf)

Installs and configures [ArchiSteamFarm](https://github.com/JustArchi/ArchiSteamFarm) on a server system.

## Requirements
### Platforms
* CentOS 7+
* Debian 8+
* Fedora 24+
* openSUSE 13+
* RHEL 7+
* Ubuntu 12.04+

## attributes
### Source Information
* `node[asf']['source']` - Root URL to where to download ArchiSteamFarm. By default, this points to the official GitHub releases page at https://github.com/JustArchi/ArchiSteamFarm/releases.
* `node['asf']['version']` - Version of ArchiSteamFarm to initially download. By default, this is `2.2.0.5`. This value might change on the system if you have chosen to auto update in your ASF global configuration file.
* `node['asf']['checksum']` - The SHA256 checksum of the `ASF.exe` for the version that you want to install.

### Install and configuration Paths
* `node['asf']['install']['path']` - Path to where to install ArchiSteamFarm. By default, this cookbook will install ArchiSteamFarm to `/opt/asf`.
* `node['asf']['config']['path']` - Path to where to store the global and bot configuration files. By default, this cookbook will place these into `/opt/asf/config`.

### Global and bot configuration
* `node['asf']['global']` - Root level key for the global ArchiSteamFarm configuration values. Look at `attributes/default.rb` for information on what values are set by default and refer to [ArchiSteamFarm's official Wiki](https://github.com/JustArchi/ArchiSteamFarm/wiki/Configuration#global-config) on information on what each setting does.

* `node['asf']['bots']`-: Root level key for all of the bot accounts that will be configured. By default, no bot account will be setup. It is necessary to set these attributes either in an environment or a wrapper/role cookbook. For information on what each value does, refer to [ArchiSteamFarm's official Wiki](https://github.com/JustArchi/ArchiSteamFarm/wiki/Configuration#bot-config). Refer to the following example attribute on how to configure a bot account called `bot_name`. It is very important that you include all of the values in the example as not setting a value will result in a broken configuration file.

```ruby
node.default['asf']['bots'] = {
  'bot_name' => {
    'AcceptConfirmationsPeriod' => 0,
    'AcceptGifts' => false,
    'CardDropsRestricted' => true,
    'CustomGamePlayedWhileFarming' => 'null',
    'CustomGamePlayedWhileIdle' => 'null',
    'DismissInventoryNotifications' => true,
    'DistributeKeys' => false,
    'Enabled' => false,
    'FarmingOrder' => 0,
    'FarmOffline' => false,
    'ForwardKeysToOtherBots' => false,
    'GamesPlayedWhileIdle' => [],
    'HandleOfflineMessages' => false,
    'IsBotAccount' => false,
    'LootableTypes' => [1, 3, 5],
    'PasswordFormat' => 0,
    'Paused' => false,
    'RedeemingPreferences' => 0,
    'SendOnFarmingFinished' => false,
    'SendTradePeriod' => 0,
    'ShutdownOnFarmingFinished' => false,
    'SteamApiKey' => 'null',
    'SteamLogin' => 'null',
    'SteamMasterClanID' => 0,
    'SteamMasterID' => 0,
    'SteamParentalPIN' => 0,
    'SteamPassword' => 'null',
    'SteamTradeToken' => 'null',
    'TradingPreferences' => 1,
  },
}
```

## Recipes
### default
The default recipe will install Mono on a target machine, install ArchiSteamFarm, place configuration files, setup a service, and then start ArchiSteamFarm.

### mono
This recipe installs Mono on a target machine. This has only been tested on the platforms outlined above in Platforms.

### service
Installs ArchiSteamFarm, loads all of configuration specified in the `global` and `bots` attributes. A service is also created for this and automatically started.

## Usage
Place a dependency on the asf cookbook in your cookbook's `metadata.rb`

```ruby
depends 'asf', '~> 1.0.0'
```

then, in a recipe:

```ruby
include_recipe 'asf::default'
```

If you would like to modify the default global configuration file or add your bots, set/override the values in `node['asf']['global']` and `node['asf']['bots']` as advised under the attributes section.

## License & Authors
- Author:: Tyler Wong ([ty-w@live.com](mailto:ty-w@live.com))

```text
Copyright 2017, Tyler Wong

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# asf Cookbook CHANGELOG

This file is used to list changes made in each version of the asf cookbook.

## 1.0.0 (2017-01-16)
- Initial release


spec/unit/recipes/mono_spec.rb
#
# Cookbook:: asf
# Spec:: mono
#
# Copyright:: 2017, Tyler Wong, All Rights Reserved.

require 'spec_helper'

describe 'asf::mono' do
  context 'Default run on ubuntu 13.10' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '13.10')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'adds the necessary apt repos' do
      expect(chef_run).to add_apt_repository('mono-wheezy')
      expect(chef_run).to add_apt_repository('mono-apache24-compat')
    end

    it 'installs the packages for ubuntu 13.10+' do
      expect(chef_run).to install_package('mono-complete')
      expect(chef_run).to install_package('referenceassemblies-pcl')
      expect(chef_run).to install_package('ca-certificates-mono')
      expect(chef_run).to install_package('mono-xsp4')
    end
  end

  context 'Default run on ubuntu 12.04' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '12.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'adds the necessary apt repos' do
      expect(chef_run).to add_apt_repository('mono-wheezy')
      expect(chef_run).to add_apt_repository('mono-wheezy-libtiff-compat')
    end

    it 'installs the packages for ubuntu 13.10+' do
      expect(chef_run).to install_package('mono-complete')
      expect(chef_run).to install_package('referenceassemblies-pcl')
      expect(chef_run).to install_package('ca-certificates-mono')
      expect(chef_run).to install_package('mono-xsp4')
    end
  end

  context 'Default run on debian 8.0' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'debian', version: '8.0')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'adds the necessary apt repos' do
      expect(chef_run).to add_apt_repository('mono-wheezy')
      expect(chef_run).to add_apt_repository('mono-libjpeg62-compat')
    end

    it 'installs the packages for debian 8.0+' do
      expect(chef_run).to install_package('mono-complete')
      expect(chef_run).to install_package('referenceassemblies-pcl')
      expect(chef_run).to install_package('ca-certificates-mono')
      expect(chef_run).to install_package('mono-xsp4')
    end
  end

  context 'Default run on rhel or fedora' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'fedora', version: '22')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'adds the yum repo' do
      expect(chef_run).to create_yum_repository('mono-project')
    end

    it 'installs the packages for rhel or fedora' do
      expect(chef_run).to install_yum_package('yum-utils')
      expect(chef_run).to install_package('mono-complete')
      expect(chef_run).to install_package('referenceassemblies-pcl')
    end
  end

  context 'Default run on suse' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'opensuse', version: '13.2')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'adds the zypper repo' do
      expect(chef_run).to add_zypper_repo('mono-project')
    end

    it 'installs the packages for suse' do
      expect(chef_run).to install_package('mono-complete')
    end
  end

  context 'Unsupported platform' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'freebsd', version: '9.3')
      runner.converge(described_recipe)
    end

    it 'converges unsuccessfully' do
      expect { chef_run }.to raise_error('Unsupported platform!')
    end
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook:: asf
# Spec:: default
#
# Copyright:: 2017, Tyler Wong, All Rights Reserved.

require 'spec_helper'

describe 'asf::default' do
  context 'When all attributes are default' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'includes the mono and service recipes' do
      expect(chef_run).to include_recipe('asf::mono')
      expect(chef_run).to include_recipe('asf::service')
    end
  end
end


spec/unit/recipes/service_spec.rb
#
# Cookbook:: asf
# Spec:: service
#
# Copyright:: 2017, Tyler Wong, All Rights Reserved.

require 'spec_helper'

describe 'asf::service' do
  context 'When all attributes are default' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04') do |node|
        node.normal['asf']['install']['path'] = '/opt/asf'
        node.normal['asf']['config']['path'] = '/opt/asf/config'
        node.normal['asf']['bots'] = {
          'bot_name' => {
            'AcceptConfirmationsPeriod' => 0,
            'AcceptGifts' => 'false',
            'CardDropsRestricted' => 'true',
            'CustomGamePlayedWhileFarming' => 'null',
            'CustomGamePlayedWhileIdle' => 'null',
            'DismissInventoryNotifications' => 'true',
            'DistributeKeys' => 'false',
            'Enabled' => 'false',
            'FarmingOrder' => 0,
            'FarmOffline' => false,
            'ForwardKeysToOtherBots' => 'false',
            'GamesPlayedWhileIdle' => [],
            'HandleOfflineMessages' => 'false',
            'IsBotAccount' => 'false',
            'LootableTypes' => [1, 3, 5],
            'PasswordFormat' => 0,
            'Paused' => 'false',
            'RedeemingPreferences' => 0,
            'SendOnFarmingFinished' => 'false',
            'SendTradePeriod' => 0,
            'ShutdownOnFarmingFinished' => 'false',
            'SteamApiKey' => 'null',
            'SteamLogin' => 'null',
            'SteamMasterClanID' => 0,
            'SteamMasterID' => 0,
            'SteamParentalPIN' => '0',
            'SteamPassword' => 'null',
            'SteamTradeToken' => 'null',
            'TradingPreferences' => 1,
          },
        }
      end.converge(described_recipe)
    end

    let(:node) { chef_run.node }

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'creates a new service user' do
      expect(chef_run).to create_poise_service_user('asf')
    end

    it 'creates the install and config directories' do
      expect(chef_run).to create_directory(node['asf']['install']['path'])
      expect(chef_run).to create_directory(node['asf']['config']['path'])
    end

    it 'downloads the asf executable' do
      expect(chef_run).to create_remote_file("#{node['asf']['install']['path']}/ASF.exe")
    end

    it 'creates the config template files' do
      expect(chef_run).to create_template("#{node['asf']['config']['path']}/ASF.json")

      node['asf']['bots'].keys.each do |bot_name|
        expect(chef_run).to create_template("#{node['asf']['config']['path']}/#{bot_name}.json")
      end
    end

    it 'creates a new service' do
      expect(chef_run).to enable_poise_service('asf')
    end
  end
end


rakefile.rb
# Style tests. cookstyle (rubocop) and Foodcritic
namespace :style do
  begin
    require 'cookstyle'
    require 'rubocop/rake_task'

    desc 'Run Ruby style checks'
    RuboCop::RakeTask.new(:ruby)
  rescue LoadError => e
    puts ">>> Gem load error: #{e}, omitting #{task.name}" unless ENV['CI']
  end

  begin
    require 'foodcritic'

    desc 'Run Chef style checks'
    FoodCritic::Rake::LintTask.new(:chef) do |t|
      t.options = {
        fail_tags: ['any'],
        progress: true,
      }
    end
  rescue LoadError
    puts ">>> Gem load error: #{e}, omitting #{task.name}" unless ENV['CI']
  end
end

desc 'Run all style checks'
task style: ['style:chef', 'style:ruby']

# ChefSpec
begin
  desc 'Run ChefSpec examples'
  require 'rspec/core/rake_task'
  RSpec::Core::RakeTask.new(:spec)
rescue LoadError => e
  puts ">>> Gem load error: #{e}, omitting #{task.name}" unless ENV['CI']
end

# Integration tests. Kitchen.ci
namespace :integration do
  begin
    require 'kitchen/rake_tasks'

    desc 'Run kitchen integration tests'
    Kitchen::RakeTasks.new
  rescue StandardError => e
    puts ">>> Gem load error: #{e}, omitting #{task.name}" unless ENV['CI']
  end
end

namespace :supermarket do
  begin
    desc 'Publish cookbook to Supermarket with Stove'
    require 'stove/rake_task'
    Stove::RakeTask.new
  rescue LoadError => e
    puts ">>> Gem load error: #{e}, omitting #{task.name}" unless ENV['CI']
  end
end

# Default
task default: %w(style spec)


recipes/service.rb
#
# Cookbook:: asf
# Recipe:: service
#
# Copyright:: 2017, Tyler Wong, All Rights Reserved.

poise_service_user 'asf' do
  action :create
end

%W(#{node['asf']['install']['path']} #{node['asf']['config']['path']}).each do |dir|
  directory dir do
    owner 'asf'
    group 'asf'
    mode '0755'
    recursive true
    action :create
  end
end

remote_file "#{node['asf']['install']['path']}/ASF.exe" do
  source File.join(node['asf']['source'], node['asf']['version'], 'ASF.exe')
  checksum node['asf']['checksum']
  owner 'asf'
  group 'asf'
  mode '0755'
  notifies :restart, 'poise_service[asf]', :delayed
  action :create
end

template "#{node['asf']['config']['path']}/ASF.json" do
  source 'ASF.erb'
  owner 'asf'
  group 'asf'
  mode '0755'
  notifies :restart, 'poise_service[asf]', :delayed
  action :create
end

node['asf']['bots'].keys.each do |bot_name|
  template "#{node['asf']['config']['path']}/#{bot_name}.json" do
    source 'bot.erb'
    owner 'asf'
    group 'asf'
    mode '0550'
    variables(bot_name: bot_name)
    notifies :restart, 'poise_service[asf]', :delayed
    action :create
  end
end

poise_service 'asf' do
  command "/usr/bin/mono #{node['asf']['install']['path']}/ASF.exe --path=#{node['asf']['install']['path']} --server"
  user 'asf'
  directory node['asf']['install']['path']
  restart_on_update true
  action :enable
end


attributes/default.rb
# Source information
default['asf']['source'] = 'https://github.com/JustArchi/ArchiSteamFarm/releases/download'
default['asf']['version'] = '2.2.0.5'
default['asf']['checksum'] = 'fc1da9028a874fb75e58e48a9d7c254580d659195f7e92dff29e157ccb7fb83e'

# Default paths
default['asf']['install']['path'] = '/opt/asf'
default['asf']['config']['path'] = '/opt/asf/config'

# Global and bot configuration values
default['asf']['global'] = {
  'AutoRestart' => true,
  'AutoUpdates' => true,
  'Blacklist' => [267_420, 303_700, 335_590, 368_020, 425_280, 480_730, 566_020],
  'Debug' => false,
  'FarmingDelay' => 15,
  'ForceHttp' => false,
  'GiftsLimiterDelay' => 1,
  'Headless' => true,
  'HttpTimeout' => 60,
  'IdleFarmingPeriod' => 3,
  'InventoryLimiterDelay' => 3,
  'LoginLimiterDelay' => 10,
  'MaxFarmingTime' => 10,
  'MaxTradeHoldDuration' => 15,
  'Statistics' => true,
  'SteamOwnerID' => 0,
  'SteamProtocol' => 6,
  'UpdateChannel' => 1,
  'WCFHost' => '127.0.0.1',
  'WCFPort' => 1242,
}

default['asf']['bots'] = {}


recipes/mono.rb
#
# Cookbook:: asf
# Recipe:: mono
#
# Copyright:: 2017, Tyler Wong, All Rights Reserved.

packages = []

case node['platform_family']
when 'debian'
  # Needed by all debian based installs
  apt_repository 'mono-wheezy' do
    uri 'http://download.mono-project.com/repo/debian'
    distribution 'wheezy'
    components ['main']
    key 'http://download.mono-project.com/repo/xamarin.gpg'
  end

  # For Ubuntu 13.10+ and Debian 8.0+
  apt_repository 'mono-apache24-compat' do
    uri 'http://download.mono-project.com/repo/debian'
    distribution 'wheezy-apache24-compat'
    components ['main']
    key 'http://download.mono-project.com/repo/xamarin.gpg'
    action :add
    only_if do
      (node['platform'] == 'ubuntu' && node['platform_version'].to_f >= 13.10) ||
        (node['platform'] == 'debian' && node['platform_version'].to_f >= 8.0)
    end
  end

  # For Debian 8.0+
  apt_repository 'mono-libjpeg62-compat' do
    uri 'http://download.mono-project.com/repo/debian'
    distribution 'wheezy-libjpeg62-compat'
    components ['main']
    key 'http://download.mono-project.com/repo/xamarin.gpg'
    action :add
    only_if { node['platform'] == 'debian' && node['platform_version'].to_f >= 8.0 }
  end

  # For Ubuntu 12.04
  apt_repository 'mono-wheezy-libtiff-compat' do
    uri 'http://download.mono-project.com/repo/debian'
    distribution 'wheezy-libtiff-compat'
    components ['main']
    key 'http://download.mono-project.com/repo/xamarin.gpg'
    action :add
    only_if { node['platform'] == 'ubuntu' && ['precise'].include?(node['lsb']['codename']) }
  end

  packages = %w(mono-complete referenceassemblies-pcl ca-certificates-mono mono-xsp4)

when 'rhel', 'fedora'
  yum_package 'yum-utils' do
    action :install
  end

  yum_repository 'mono-project' do
    description 'Mono yum repo'
    baseurl 'http://download.mono-project.com/repo/centos/'
    gpgkey 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF'
    action :create
  end

  packages = %w(mono-complete referenceassemblies-pcl)

when 'suse'
  zypper_repo 'mono-project' do
    uri 'http://download.mono-project.com/repo/centos/'
    key 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF'
    action :add
  end

  packages = %w(mono-complete)

else
  raise 'Unsupported platform!'
end

packages.each do |pkg|
  package pkg do
    action :install
  end
end
